In [1]:
!pip install graphdatascience

In [2]:
!pip install emblaze

In [3]:
from graphdatascience import GraphDataScience
import pandas as pd
import numpy as np
import emblaze

Before running this notebook, set up a free Neo4j sandbox at https://sandbox.neo4j.com/.

Choose the data science project when you create your sandbox.

Copy the Bolt URL and password from your sandbox connection information into the cell below.

In [4]:
bolt_url = "bolt://44.198.160.170:7687"
user = "neo4j"
password = "specifications-lifeboats-jacket"
gds = GraphDataScience(bolt_url, auth=(user, password))

Add a weight property to the HAS_ROUTE relationships that will have a higher value for airports that are geographically closer together.

In [5]:
gds.run_cypher("""
    match (:Airport)-[r:HAS_ROUTE]->(:Airport)
    with collect(r) as routes, max(r.distance) as maxDistance
    foreach(route in routes | set route.weight = maxDistance + 1 - route.distance)
""")

Empty DataFrame
Columns: []
Index: []

Project an in-memory graph containing Airport nodes connected by HAS_ROUTE relationships.

In [6]:
G_routes, result = gds.graph.project("air-routes", "Airport", 
                                     {"HAS_ROUTE":{"orientation":"UNDIRECTED",
                                                   "aggregation":"MAX"}}, 
                                      relationshipProperties = "weight")

In [7]:
result

nodeProjection            {'Airport': {'label': 'Airport', 'properties':...
relationshipProjection    {'HAS_ROUTE': {'orientation': 'UNDIRECTED', 'a...
graphName                                                        air-routes
nodeCount                                                              3503
relationshipCount                                                     49476
projectMillis                                                           303
Name: 0, dtype: object

Verify that the HAS_ROUTE relationships in the in-memory graph have a "weight" property.

In [8]:
G_routes.relationship_properties("HAS_ROUTE")

['weight']

Let's see how many weakly connected components are in the graph.

In [9]:
routes_wcc = gds.wcc.stats(G_routes)
routes_wcc

componentCount                                                         212
componentDistribution    {'p99': 1, 'min': 1, 'max': 3292, 'mean': 16.5...
postProcessingMillis                                                     2
preProcessingMillis                                                      0
computeMillis                                                            4
configuration            {'seedProperty': None, 'consecutiveIds': False...
Name: 0, dtype: object

In [10]:
routes_wcc['componentDistribution']

{'p99': 1,
 'min': 1,
 'max': 3292,
 'mean': 16.52358490566038,
 'p90': 1,
 'p50': 1,
 'p999': 3292,
 'p95': 1,
 'p75': 1}

It looks like we have a lot of isolated nodes (component size 1.) Add a componentId property for the wcc component to the in-memory graph.

In [11]:
gds.wcc.mutate(G_routes, mutateProperty = 'componentId')

mutateMillis                                                             0
nodePropertiesWritten                                                 3503
componentCount                                                         212
componentDistribution    {'p99': 1, 'min': 1, 'max': 3292, 'mean': 16.5...
postProcessingMillis                                                     2
preProcessingMillis                                                      0
computeMillis                                                            4
configuration            {'seedProperty': None, 'consecutiveIds': False...
Name: 0, dtype: object

Write the in-memory property out to the graph on disk.

In [12]:
gds.graph.writeNodeProperties(G_routes, ['componentId'])

writeMillis                     24
graphName               air-routes
nodeProperties       [componentId]
propertiesWritten             3503
Name: 0, dtype: object

Get the componentId that contains most of the nodes (the giant component).

In [13]:
gds.run_cypher("MATCH (a:Airport) RETURN a.componentId as componentId, count(*) as nodeCount ORDER BY count(*) DESC limit 1")

componentId  nodeCount
0            0       3292

Project a subgraph that contains only Airports in the giant component.

In [14]:
G_connected_airports, result = gds.beta.graph.project.subgraph("connected-airports", G_routes, "n.componentId = 0", "*")

Confirm that the number of nodes in the subgraph is 3,292, the same as in the maximum component size from the wcc stats.

In [15]:
G_connected_airports.node_count()

3292

Write a function that adds a fastRP embedding of dimension 64 with some keyward arguments to the in-memory graph.

In [16]:
def train_fast_rp(graph, config):
    result = gds.fastRP.mutate(
        graph,
        embeddingDimension = 64,
        randomSeed = 45,
        **config
    )
    return result

Experiment with different values for the [iterationWeights property](https://neo4j.com/docs/graph-data-science/current/machine-learning/node-embeddings/fastrp/). The longer the list of values for this property, the more distantly related nodes can influence the embedding.

In [17]:
configs = [{"iterationWeights": [1.0, 1.0], "mutateProperty": "shallowUnweighted"},
           {"iterationWeights": [0.0, 1.0, 1.0], "mutateProperty": "mediumUnweighted"},
           {"iterationWeights": [1.0, 1.0, 1.0, 1.0], "mutateProperty": "deepUnweighted"},
           {"iterationWeights": [1.0, 1.0], "relationshipWeightProperty": "weight", 
            "mutateProperty": "shallowWeighted"},
           {"iterationWeights": [0.0, 1.0, 1.0], "relationshipWeightProperty": "weight", 
            "mutateProperty": "mediumWeighted"},
           {"iterationWeights": [1.0, 1.0, 1.0, 1.0], "relationshipWeightProperty": "weight", 
            "mutateProperty": "deepWeighted"}]                      

Call the embedding for each configuration and review the results.

In [18]:
embedding_results = [train_fast_rp(G_connected_airports, config) for config in configs]

In [19]:
pd.DataFrame(embedding_results)

nodePropertiesWritten  mutateMillis  nodeCount  preProcessingMillis  \
0                   3292             0       3292                    0   
0                   3292             0       3292                    0   
0                   3292             0       3292                    0   
0                   3292             0       3292                    0   
0                   3292             0       3292                    0   
0                   3292             0       3292                    0   

   computeMillis                                      configuration  
0             14  {'nodeSelfInfluence': 0, 'relationshipWeightPr...  
0             19  {'nodeSelfInfluence': 0, 'relationshipWeightPr...  
0             22  {'nodeSelfInfluence': 0, 'relationshipWeightPr...  
0            127  {'nodeSelfInfluence': 0, 'relationshipWeightPr...  
0            256  {'nodeSelfInfluence': 0, 'relationshipWeightPr...  
0            272  {'nodeSelfInfluence': 0, 'relationshipWeightPr...

Stream the embedding values back from the in-memory graph and look up the airport description and continent.

In [20]:
embedding_df = gds.run_cypher("""
    call gds.graph.streamNodeProperties("connected-airports", 
        ["shallowUnweighted", "mediumUnweighted", "deepUnweighted", "shallowWeighted", "mediumWeighted", "deepWeighted"]) 
    yield nodeId, nodeProperty, propertyValue
    WITH gds.util.asNode(nodeId) as a,
    MAX(case when nodeProperty = "shallowUnweighted" then propertyValue end) as shallowUnweighted,
    MAX(case when nodeProperty = "mediumUnweighted" then propertyValue end) as mediumUnweighted,
    MAX(case when nodeProperty = "deepUnweighted" then propertyValue end) as deepUnweighted,
    MAX(case when nodeProperty = "shallowWeighted" then propertyValue end) as shallowWeighted,
    MAX(case when nodeProperty = "mediumWeighted" then propertyValue end) as mediumWeighted,
    MAX(case when nodeProperty = "deepWeighted" then propertyValue end) as deepWeighted
    MATCH (a)-[:ON_CONTINENT]->(c:Continent)
    RETURN
    a.descr as airport_name, 
    a.iata as airport_code, 
    c.name as continent,
    shallowUnweighted,
    mediumUnweighted,
    deepUnweighted,
    shallowWeighted,
    mediumWeighted,
    deepWeighted
    ORDER BY size([(a)-[:HAS_ROUTE]-() | a]) DESC
    LIMIT 900
    """)

In [21]:
embedding_df

airport_name airport_code continent  \
0                  Frankfurt am Main          FRA        EU   
1            Paris Charles de Gaulle          CDG        EU   
2     Istanbul International Airport          IST        EU   
3         Amsterdam Airport Schiphol          AMS        EU   
4       Munich International Airport          MUC        EU   
..                               ...          ...       ...   
895         Nanping Wuyishan Airport          WUS        AS   
896     Scott AFB/Midamerica Airport          BLV        NA   
897   Canberra International Airport          CBR        OC   
898  Mangalore International Airport          IXE        AS   
899           Cherry Capital Airport          TVC        NA   

                                     shallowUnweighted  \
0    [-0.07399057596921921, 0.2620871663093567, 0.1...   
1    [-0.06564060598611832, 0.055160876363515854, 0...   
2    [0.026787318289279938, 0.2438107430934906, 0.1...   
3    [-0.1855502426624298, 0.16803550720214844, 0.1...   
4    [-0.20580337941646576, 0.11519283056259155, 0....   
..                                                 ...   
895  [-0.15465746819972992, -0.29468223452568054, -...   
896  [0.3646937608718872, 0.18155963718891144, -0.1...   
897  [0.0862639844417572, 0.5978029370307922, 0.028...   
898  [0.0036581531167030334, 0.37329620122909546, 0...   
899  [0.14239120483398438, 0.07686980068683624, 0.2...   

                                      mediumUnweighted  \
0    [-0.11316736042499542, 0.24514061212539673, 0....   
1    [-0.11409448832273483, 0.21923694014549255, 0....   
2    [-0.025942979380488396, 0.2735888957977295, 0....   
3    [-0.16270774602890015, 0.2214188426733017, 0.1...   
4    [-0.16538169980049133, 0.2382025420665741, 0.1...   
..                                                 ...   
895  [-0.16766786575317383, -0.05741892755031586, -...   
896  [0.38825953006744385, 0.09397226572036743, -0....   
897  [0.18571466207504272, 0.6202242374420166, 0.07...   
898  [0.11747836321592331, 0.337380051612854, 0.131...   
899  [0.2795731723308563, 0.05664905160665512, 0.08...   

                                        deepUnweighted  \
0    [-0.15192122757434845, 0.5060424208641052, 0.2...   
1    [-0.15566448867321014, 0.28849947452545166, 0....   
2    [-0.0012168977409601212, 0.5254808664321899, 0...   
3    [-0.297057569026947, 0.40379035472869873, 0.24...   
4    [-0.3348800837993622, 0.3535888195037842, 0.32...   
..                                                 ...   
895  [-0.28116992115974426, -0.25400516390800476, -...   
896  [0.6421867609024048, 0.2155066579580307, -0.12...   
897  [0.1772807538509369, 1.2233781814575195, 0.106...   
898  [0.08971487730741501, 0.7435741424560547, 0.12...   
899  [0.36861124634742737, 0.10893549770116806, 0.2...   

                                       shallowWeighted  \
0    [-0.19411295652389526, 0.2675624191761017, 0.2...   
1    [-0.15881961584091187, 0.0621316060423851, 0.2...   
2    [-0.0003005862236022949, 0.26281487941741943, ...   
3    [-0.28153276443481445, 0.19925662875175476, 0....   
4    [-0.29063165187835693, 0.13292333483695984, 0....   
..                                                 ...   
895  [-0.16491110622882843, -0.310563325881958, -0....   
896  [0.3598550856113434, 0.179595947265625, -0.126...   
897  [0.04642222821712494, 0.528540849685669, 0.041...   
898  [-0.029092632234096527, 0.34626173973083496, 0...   
899  [0.13504764437675476, 0.0393684059381485, 0.25...   

                                        mediumWeighted  \
0    [-0.1977788209915161, 0.2508077621459961, 0.13...   
1    [-0.20260721445083618, 0.22906702756881714, 0....   
2    [-0.06701691448688507, 0.2827916741371155, 0.1...   
3    [-0.2414441704750061, 0.22703152894973755, 0.1...   
4    [-0.233033686876297, 0.24436715245246887, 0.15...   
..                                                 ...   
895  [-0.19777408242225647, -0.0949367955327034, -0...   
896  [0.3890777

Create Emblaze embeddings for each graph embedding. Compute the nearest neighbors and create a 2-dimensional projection using UMAP.

In [28]:
def create_emblaze_embedding(embedding_df, column):
    emb = emblaze.Embedding({emblaze.Field.POSITION: np.array(list(embedding_df[column])),
                             emblaze.Field.COLOR: embedding_df['continent']}, 
                             n_neighbors = 10,
                             label=column, 
                             metric='cosine')
    emb.compute_neighbors()
    return emb.project()

In [29]:
emblaze_embeddings = [create_emblaze_embedding(embedding_df, column) for column in embedding_df.columns[3:]]

Turn the list of embeddings into an embedding set.

In [30]:
variants = emblaze.EmbeddingSet(emblaze_embeddings)

Create descriptions for each point in the visualization.

In [31]:
thumbnails = emblaze.TextsThumbnails(embedding_df['airport_name'] + " (" + embedding_df['airport_code'] + ")")

Create the emblaze widget and display it.

In [32]:
w = emblaze.Viewer(embeddings = variants, thumbnails = thumbnails)

In [33]:
w

Viewer(colorScheme='tableau', data={'data': [{'_format': 'compressed', '_idtype': 'u2', '_length': 900, 'ids':…

Clean up the in-memory graphs.

In [39]:
G_connected_airports.drop()
G_routes.drop()